lib/Preview/EPubPreview.php

```diff
@@ -0,0 +1,245 @@
+						$references = $package->guide->children();
+						foreach($references as $reference) {
+							if ($reference['type'] == 'cover' || $reference['type'] == 'title-page') {
+								$cover = (string) $reference['href'];
+								break;
+							}
+						}
+					}
+
+					// no cover ? no image ? take the first page
+					if (!$img_src && !$cover) {
+						$first_page_id = (string) $package->spine->itemref['idref'];
+						if ($first_page_id) {
+							foreach($items as $item) {
+								if ($item['id'] == $first_page_id) {
+									$cover = (string) $item['href'];
+									break;
+								}
+							}
+
+						}
+					}
+
+					// have we a "cover" file ?
+					if ($cover) {
+						// relative to container
+						$img_src = null;
+						$path = $this->resolvePath($path, $cover);
+						$dom = $this->extractHTML($path);
+						if ($dom) {
+							// search img
+							$images = $dom->getElementsByTagName('img');
+							if ($images->length) {
+								$img_src = $images[0]->getAttribute('src');
+							} else {
+								$images = $dom->getElementsByTagName('image');
+								if ($images->length) {
+									$img_src = $images[0]->getAttribute('xlink:href');
+								}
+							}
+						}
+					}// cover
+
+					// img ?
+					if ($img_src) {
+						$img_src = $this->resolvePath($path, $img_src);
+						$img_data = $this->extractFileData($img_src);
+					}
+				}
+			}
+
+			// Pfff. Make a pause
+			if ($img_data) {
+				$image = new \OC_Image();
+				$image->loadFromData($img_data);
+				return $image;
+			}
+			return null;
+		} catch (\Exception $e) {
+			return null;
+		}
+	}
+
+	/**
+	 * find the main content XML (usually "content.opf")
+	 */
+	private function getContentPath() : ?string {
+		$xml_container = $this->extractXML('META-INF/container.xml');
+		if (is_object($xml_container)) {
+			$full_path = $xml_container->rootfiles->rootfile['full-path'][0];
+			if ($full_path) {
+				return $full_path->__toString();
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * extract HTML from Zip path
+	 * @param string $path
+	 * @return \DOMDocument|null
+	 */
+	protected function extractHTML(string $path): \DOMDocument|null {
+		$html = $this->extractFileData($path);
+		if (is_string($html)) {
+			$dom = new \DOMDocument('1.0', 'utf-8');
+			$dom->strictErrorChecking = false;
+			if (@$dom->loadHTML($html)) {
+				return $dom;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * extract XML from Zip path
+	 *
+	 * @psalm-param 'META-INF/container.xml' $path
+	 */
+	private function extractXML(string $path): \SimpleXMLElement|false|null {
+		$xml = $this->extractFileData($path);
+		if (is_string($xml)) {
+			return simplexml_load_string($xml);
+		}
+		return null;
+	}
+
+	/**
+	 * get unzipped data
+	 *
+	 * @param string $path file path in zip
+	 *
+	 * @psalm-param 'META-INF/container.xml' $path
+	 *
+	 * @return false|null|string
+	 */
+	private function extractFileData(string $path): string|false|null {
+		$fp = $this->zip->getStream($path, 'r');
```

Comment:
@smarinier Does `OC\Archive\ZIP` exist at all if the app `files_zip` isn't installed?

Reply:
OK so I guess it's a private API: https://github.com/nextcloud/server/blob/952271929d888c8333f5b64aa676f802a8b682af/lib/private/Archive/ZIP.php#L13

How much effort would it require to change the code to use PHP's [`ZipArchive`](https://www.php.net/manual/en/class.ziparchive.php)?

Reply:
I think this would be a pity, as the ZIP class embeds ZipArchive properly, and logs errors in NextCloud. Its probably a definition that is necessary to do with Psalm.

Reply:
The problem is that it's a *private* file that should *not* be used in apps. This means that if we use this implementation it will unexpectedly break in an update without any warnings. I don't want to maintain code that relies on non-standard internal APIs.